### PR TITLE
SBS add jwt support

### PR DESCRIPTION
**Pull request by petro2061**
Added new clasees: JwtUtil - incude methods for generat JWT token and that validation, JwtAuthenticationFilter - implemented custom filter that checked corect authentication users bases on JWT token, AuthenticationService - implemented logic for authentication users. Also added swagger security and corresponding annotations. Also added necessary changes to configuration files, classes for implementing stateless applications.

### CODE REVIEW

File: src/main/java/project/onlinebookstore/config/OpenApiConfig.java [hunk 1]
```diff
@@ -13,6 +16,12 @@ public OpenAPI customOpenApi() {
...
                         .title("Online Book Store API")
                         .version("1.0.0")
-                        .description("API for management online book store"));
+                        .description("API for management online book store"))
+                .components(new Components().addSecuritySchemes("BearerAuth",
+                        new SecurityScheme()
+                                .type(SecurityScheme.Type.HTTP)
+                                .scheme("bearer")
+                                .bearerFormat("JWT")))
+                .addSecurityItem(new SecurityRequirement().addList("BearerAuth"));
```
Comment (MentorMateAcademy) [inline src/main/java/project/onlinebookstore/config/OpenApiConfig.java]: strings can save to constants

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

**Comment by petro2061** [comment]
Added constsnts and lose anntations

File: src/main/java/project/onlinebookstore/config/OpenApiConfig.java [hunk 2]
```diff
@@ -1,18 +1,31 @@
...
 import io.swagger.v3.oas.models.OpenAPI;
 import io.swagger.v3.oas.models.info.Info;
+import io.swagger.v3.oas.models.security.SecurityRequirement;
+import io.swagger.v3.oas.models.security.SecurityScheme;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
 @Configuration
 public class OpenApiConfig {
+    private static final String SCHEMA_TYPE = "bearer";
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/config/OpenApiConfig.java]: why do you need two similar constant?

File: src/main/java/project/onlinebookstore/config/OpenApiConfig.java [hunk 3]
```diff
@@ -1,18 +1,31 @@
...
+    private static final String BEARER_FORMAT = "bearer";
+    private static final String SECURITY_ITEM_LIST_NAME = "BearerAuth";
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/config/OpenApiConfig.java]:
```suggestion
    private static final String SECURITY_SCHEME_NAME = "BearerAuth";
```

File: src/main/java/project/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 4]
```diff
@@ -55,6 +56,22 @@ public ResponseEntity<Map<String, Object>> handleAccessDeniedException(
                 HttpStatus.FORBIDDEN);
     }
 
+    @ExceptionHandler(JwtException.class)
+    public ResponseEntity<Map<String, Object>> handleJwtException(
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/exception/CustomGlobalExceptionHandler.java]: check this, GlogalExceptionHandler doesn't handle exceptions from filters

File: src/main/java/project/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+@RequiredArgsConstructor
+@Component
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private static final String HEADER_NAME = "Authorization";
+    private static final String AUTH_TYPE = "Bearer ";
+    private static final int START_TOKEN_INDEX = 7;
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
    private static final String TOKEN_HEADER = "Bearer ";
```

File: src/main/java/project/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                    userDetails, null, userDetails.getAuthorities());
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String barerToken = request.getHeader(HEADER_NAME);
+        if (StringUtils.hasText(barerToken) && barerToken.startsWith(AUTH_TYPE)) {
+            return barerToken.substring(START_TOKEN_INDEX);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
            return barerToken.substring(TOKEN_HEADER.length());
```

File: src/main/java/project/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            UserDetails userDetails = userDetailsService.loadUserByUsername(username);
+            Authentication authentication = new UsernamePasswordAuthenticationToken(
+                    userDetails, null, userDetails.getAuthorities());
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String barerToken = request.getHeader(HEADER_NAME);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
        String barerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
```

use Spring constant

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Comment by petro2061** [comment]
Mentor comments corrected. Deleted exseption handlers in GlobalExseptionHandler.

File: src/main/java/project/onlinebookstore/config/OpenApiConfig.java [hunk 8]
```diff
@@ -1,18 +1,30 @@
...
+
     @Bean
     public OpenAPI customOpenApi() {
         return new OpenAPI()
                 .info(new Info()
                         .title("Online Book Store API")
                         .version("1.0.0")
-                        .description("API for management online book store"));
+                        .description("API for management online book store"))
+                .components(new Components().addSecuritySchemes("BearerAuth",
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/config/OpenApiConfig.java]:
```suggestion
                .components(new Components().addSecuritySchemes(SECURITY_SCHEMA_NAME,
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job. Just one comment

**Comment by petro2061** [comment]
added lose annotation to OpenApi config

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**